File: LinkedLists/Java/DetectLoop.java
package LinkedLists.Java;

public class DetectLoop {
    public static class Node {
        int data;
        Node next;

        Node(int data) {
            this.data = data;
            this.next = null;
        }
    }

    private static Node convertArrayToLinkedList(int[] arr) {
        if (arr.length == 0) return null;
        Node head = new Node(arr[0]);
        Node current = head;            
        for (int i = 1; i < arr.length; i++) {
            current.next = new Node(arr[i]);
            current = current.next;
        }
        return head;
    }

    public static boolean detectLoopHashing(Node head) {
        // Time Complexity: O(N) where N is the number of nodes in the linked list
        // Space Complexity: O(N) since we are using a HashSet to store visited nodes
        java.util.HashSet<Node> visitedNodes = new java.util.HashSet<>();
        Node current = head;

        while (current != null) {
            if (visitedNodes.contains(current)) {
                System.out.println("Loop detected at node with data: " + current.data);
                return true; // Loop detected
            }
            visitedNodes.add(current);
            current = current.next;
        }
        System.out.println("No loop detected");
        return false; // No loop detected
    }

    public static boolean detectLoop(Node head) {
        // Time Complexity: O(N) where N is the number of nodes in the linked list
        // Space Complexity: O(1) since we are using two pointers only
        if (head == null || head.next == null) return false;

        Node slow = head;
        Node fast = head;

        while (fast != null && fast.next != null) {
            slow = slow.next; // Move slow pointer by 1
            fast = fast.next.next; // Move fast pointer by 2

            if (slow == fast) {
                return true; // Loop detected
            }
        }
        return false; // No loop detected
    }

    public static void main(String[] args) {
        Node head = convertArrayToLinkedList(new int[]{1, 2, 3, 4, 5, 6,7,8,9});
        // Creating a loop for testing
        head.next.next.next.next.next = head.next.next; // 5 -> 3 creates a loop
        boolean hasLoop = detectLoop(head);
        System.out.println("Does the linked list have a loop? " + hasLoop);
        boolean hasLoopHashing = detectLoopHashing(head);
        System.out.println("Does the linked list have a loop (using hashing)? " + hasLoopHashing);
    }
}

File: LinkedLists/Python/detectLoop.py
class Node:
    def __init__(self,data1,next1=None):
        self.data = data1
        self.next = next1

def loopBrute(head):
    st = set()
    tmp = head
    
    while tmp:
        if tmp in st:
            return True
        st.add(tmp)
        tmp = tmp.next
    
    return False
# Since the distance b/n slow and fast is bound to be 1 in a loop
# We move fast by 2 nodes
def loopOptimal(head):
    slow = fast = head

    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow==fast:
            return True

    return False

def printLL(head):
    while head:
        print(head.data,end=' ')
        head = head.next
    print()

def main():
    head = Node(1)
    second = Node(2)
    third = Node(3)
    fourth = Node(4)
    fifth = Node(5)

    head.next = second
    second.next = third
    third.next = fourth
    fourth.next = fifth
    fifth.next = third
    loop1 = loopBrute(head)
    print('Does loop exists using brute force? ',loop1)
    loop2 = loopOptimal(head)
    print('Does loop exists using tortoise hare? ',loop2)

if __name__ == "__main__":
    main()